python/cudf/cudf/testing/testing.py

```diff
@@ -203,7 +204,11 @@ def assert_column_equal(
 
     columns_equal = False
     try:
-        columns_equal = left.equals(right)
+        columns_equal = (
+            left.equals(right)
+            if check_exact
+            else cp.allclose(left.values, right.values, rtol=rtol, atol=atol)
```

Comment:
Performing `.values` on categoricals is causing a failure.

Reply:
Looks like it's supposed to decategorize them in the logic above but it possibly should be `if check_dtype` ?

Reply:
(how) does this work when there are nulls in the columns?

Reply:
> Performing `.values` on categoricals is causing a failure.
> Looks like it's supposed to decategorize them in the logic above but it possibly should be `if check_dtype` ?

I think `check_exact` is the correct boolean to verify here, I think the problem is just that approximate comparison only makes sense for numerical types. For strings, categoricals, and maybe even nested types, should we just ignore the flag (and the atol/rtol args)? I assume that's how pandas deals with non-numerical types.

> (how) does this work when there are nulls in the columns?
Good catch. I assumed that calling `.values` would mask NULL with None, I didn't realize that `cupy` wouldn't support it. In that case we have to include slightly more complex logic here. Perhaps something like
```
cp.all(left.isnull() == right.isnull()) and cp.allclose(left[~left.isnull()].values, right[~right.isnull()].values)
```

Reply:
We have `NULL_EQUALS` plumbed up into the column classes if that helps, that's basically `eq` where `NA == NA` and `(NA == NOT_NA) == False`.

Reply:
What I would expect if we're doing an "approximately equal" test is to use null equality for comparing nulls but then use something like `np.allclose` for values that are present. I don't think columns should directly support something like `allclose`, that feels out of scope and pandas only provides this in the form of the corresponding `pandas.testing` APIs. So the question is basically how to do the almost equal comparison for non-null values, and I think `cupy.allclose` on the underlying `values` is the easiest option.

Reply:
If tolerance doesn't really make sense for non-numeric types should we just raise in those cases?

Reply:
That would be a nice out, but we default to `check_exact=False` (matching pandas) and I don't think it makes sense to raise an Exception if a user tries to call this API with non-numeric inputs just because they didn't modify the default.